### PR TITLE
SubwayService and SubwayServiceTest classes

### PR DESCRIPTION
**Pull request by R-Munitz**

### CODE REVIEW

File: src/main/java/touro/subwaymap/StationNode.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        this.station = station;
+        unvisited = true;
+        distance = Integer.MAX_VALUE;
+        previousStation = null;
+        adjacentStations = station.properties.adjacentStations;
+    }
+
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (andrewoid) [inline src/main/java/touro/subwaymap/StationNode.java:26]: The only thing you care about with equals and hashcode is the Station object

File: src/main/java/touro/subwaymap/SubwayMap.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.Stack;
+
+import static touro.subwaymap.SubwayStations.*;
+
+public class SubwayMap {
+    //map station objects to station nodes?
+    private final Map<Station,StationNode> subwayMap = new HashMap<>();
+    final int length = 1;
+
+    public void createMap(SubwayStations subwayStations){
```
Comment (andrewoid) [inline src/main/java/touro/subwaymap/SubwayMap.java:15]: This should be a constructor.

File: src/main/java/touro/subwaymap/SubwayMap.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            //get station with min distance
+            StationNode closestStationNode = null;
+            int minDistance = Integer.MAX_VALUE;
+            for (StationNode stationNode : unvisitedStationNodes){
+                int distance = stationNode.distance;
+                if (distance < minDistance){
+                    minDistance = distance;
+                    closestStationNode = stationNode;
+                }
+            }
```
Comment (andrewoid) [inline src/main/java/touro/subwaymap/SubwayMap.java:50]: move this to a method that returns a StationNode

File: src/main/java/touro/subwaymap/SubwayMap.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        }
+        Stack <StationNode> stationNodesStack = new Stack<>();
+        Station currentStation = end;
+        StationNode currentStationNode = endNode;
+        while (!currentStation.equals(start)){ //x compare nodes, compare station objects
+            stationNodesStack.add(currentStationNode);
+            currentStationNode = currentStationNode.previousStation;
+
+        }
```
Comment (andrewoid) [inline src/main/java/touro/subwaymap/SubwayMap.java:74]: move this into it's own method that returns a List of Stations.

getShortestRoute() should return it.

**Comment by andrewoid** [comment]
What happened to the rest of your code in this PR?

Contact me immediately.

**Comment by Rmunitz-oss** [comment]
Just saw your comment. Deleted those classes and created a different one instead with a working algorithm, was not ready to push it earlier.

File: src/main/java/touro/subwaymap/SubwayService.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    private Station start;
+    private Station end;
+    private ArrayList<Station> unvisitedStations = new ArrayList<>();
+    private Station closestStation;
+    private final int length = 1;
+
+    //where should this kind of setup be?
+    // if not here,getSubwayRoute would need a SubwayStations object as a param? (not intuitive)
+    //create constructor for SubwayService that takes SubwayStation arg?
+    public void initialize() throws IOException {
```
Comment (andrewoid) [inline src/main/java/touro/subwaymap/SubwayService.java:20]: This  should be in a constructor

File: src/main/java/touro/subwaymap/SubwayService.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
+package touro.subwaymap;
+
+import java.io.IOException;
+import java.util.*;
+import static java.util.Collections.*;
+import static touro.subwaymap.SubwayStations.*;
+
+public class SubwayService {
+    //clean, which belong in methods? public vs private?
```
Comment (andrewoid) [inline src/main/java/touro/subwaymap/SubwayService.java:9]: if a variable is only used in a method, it should be declared in that method. Not as a member variable.

File: src/main/java/touro/subwaymap/SubwayService.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    //create constructor for SubwayService that takes SubwayStation arg?
+    public void initialize() throws IOException {
+        JSONReader jsonReader = new JSONReader();
+        subwayStations = jsonReader.readSubwayStationJSON();
+        Map<String, ArrayList<Integer>> lineStationMap = jsonReader.readSubwayLinesJSON();
+        SubwayDataManager subwayDataManager = new SubwayDataManager();
+        subwayDataManager.processSubwayData(subwayStations, lineStationMap);
+    }
+
+    public ArrayList<Station> getSubwayRoute (double [] startLocation, double [] endLocation) {//what params would really be passed?
```
Comment (andrewoid) [inline src/main/java/touro/subwaymap/SubwayService.java:28]: this is fine

File: src/main/java/touro/subwaymap/SubwayService.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        start = findNearestStation(startLocation[0],startLocation[1]);
+        end = findNearestStation(endLocation[0],endLocation[1]);
+        ArrayList<Station> shortestRoute = findShortestRoute();
+        return shortestRoute; //append final station? rewrite loop to include end/ exclude start?
+    }
+
+    private Station findNearestStation(double longitude, double latitude) {
```
Comment (andrewoid) [inline src/main/java/touro/subwaymap/SubwayService.java:35]: usually it's latitude, longitude

Comment (Rmunitz-oss) [reply src/main/java/touro/subwaymap/SubwayService.java:35]: In the json file, it gives the longitude of each station first. Should I still change my methods to match convention?

File: src/main/java/touro/subwaymap/SubwayService.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+    private ArrayList<Station> getStationPath() {
+        ArrayList <Station> stationPath = new ArrayList<>();
+        Station currentStation = end;
+        stationPath.add(currentStation);
+        while (!currentStation.equals(start)){
+            stationPath.add(currentStation.properties.previousStation);
+            currentStation = currentStation.properties.previousStation;
+        }
+        reverse(stationPath);
```
Comment (andrewoid) [inline src/main/java/touro/subwaymap/SubwayService.java:86]: Make sure you add the start station

File: src/test/java/touro/subwaymap/SubwayServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        expectedShortestRoute.add(subwayStations.stations.get(357)); //34th St - Penn Station
+        expectedShortestRoute.add(subwayStations.stations.get(358)); //Times Sq - 42nd St
+        expectedShortestRoute.add(subwayStations.stations.get(2));   //50th St
+        expectedShortestRoute.add(subwayStations.stations.get(93));  //59th St - Columbus Circle (end)
+
+        //when
+        ArrayList<Station> shortestRoute = subwayService.getSubwayRoute(startLocation,endLocation);
+
+        //then
+        assertEquals(expectedShortestRoute,shortestRoute);
```
Comment (andrewoid) [inline src/test/java/touro/subwaymap/SubwayServiceTest.java:37]: 👍

**Review by andrewoid — APPROVED** [review]

File: src/main/java/touro/subwaymap/SubwayService.java (hunk 9 above)
Comment (Rmunitz-oss) [reply src/main/java/touro/subwaymap/SubwayService.java:86]: It is already added when the `currentStation.properties.previousStation` is added to the path

Comment (andrewoid) [reply src/main/java/touro/subwaymap/SubwayService.java:86]: oooooh. Read it wrong.

**Review by andrewoid — APPROVED** [review]

**Pull request merged**